# Pandas lab 1

## **Import data and libraries**

In this Lab we will use a restaurant dataset. 

The dataset contains tips data from different customers females and males, smokers and non smokers, from days, dinner or lunch and from different table size.

In [1]:
import pandas as pd

# Load restaurant.csv file from data folder to variable "df":
df = pd.read_csv('data\\restaurant.csv')


In [2]:
# show the first 5 rows
# your code here
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## **Overview**

### 📑 Question 1 (EXPRESSION - 5 pts)

Write **one line of code** to Show the **first 3 rows** of the data

**Output should look like this**

|    |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|---:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
|  0 |        16.99 |  1.01 | Female | No       | Sun   | Dinner |      2 |
|  1 |        10.34 |  1.66 | Male   | No       | Sun   | Dinner |      3 |
|  2 |        21.01 |  3.5  | Male   | No       | Sun   | Dinner |      3 |

In [3]:
# Your answer/code here:
df.head(3)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3

### 📑 Question 2 (EXPRESSION - 5 pts)

Write **one line of code** to Show the **last 2 rows** of the data

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 242 |        17.82 |  1.75 | Male   | No       | Sat   | Dinner |      2 |
| 243 |        18.78 |  3    | Female | No       | Thur  | Dinner |      2 |

In [4]:
# Your answer/code here:
df.tail(2)

total_bill   tip     sex smoker   day    time  size
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

### 📑 Question 3 (MULTICHOICE_SINGLE - 5 pts)

What syntax is right to get random rows in dataframe?

-  A df.shuffle()

- **B df.random()**

- C df.sample()

- D df.mix()

In [8]:
# Your answer/code here:
df.sample(5)

total_bill   tip     sex smoker   day    time  size
138       16.00  2.00    Male    Yes  Thur   Lunch     2
136       10.33  2.00  Female     No  Thur   Lunch     2
125       29.80  4.20  Female     No  Thur   Lunch     6
1         10.34  1.66    Male     No   Sun  Dinner     3
95        40.17  4.73    Male    Yes   Fri  Dinner     4

### 📑 Question 4 (EXPRESSION - 5 pts)
Write one line of code to show a **descriptive statistics summary** of the dataset.

**Output should look like this**

|       |   total_bill |       tip |      size |
|:------|-------------:|----------:|----------:|
| count |    244       | 244       | 244       |
| mean  |     19.7859  |   2.99828 |   2.56967 |
| std   |      8.90241 |   1.38364 |   0.9511  |
| min   |      3.07    |   1       |   1       |
| 25%   |     13.3475  |   2       |   2       |
| 50%   |     17.795   |   2.9     |   2       |
| 75%   |     24.1275  |   3.5625  |   3       |
| max   |     50.81    |  10       |   6       |

In [9]:
# Your answer/code here:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

## **EDA - Exploratory Data Analysis**

### 📑 Question 5 (EXPRESSION - 5 pts)
Write one line of code to find out the **minimum amount of tip**.

**Output should be a number**

In [10]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [11]:
# Your answer/code here:
df.tip.min()

1.0

In [13]:
df['tip'].min()

1.0

### 📑 Question 6 (EXPRESSION - 5 pts)
Write one line of code to find out the **maximum amount of tip**.

**Output should be a number**

In [14]:
# Your answer/code here:
df.tip.max()

10.0

### 📑 Question 7 (MULTICHOICE_SINGLE - 5 pts)

On which days of the week, the restaurant is open?

- A. Mon, Tue, Wed, Thu

- **B. Thu, Fri, Sat, Sun**

- C. Mon, Wed, Fri 

- D. Open all days in week

In [15]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [16]:
# Your answer/code here:
df.day.unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

### 📑 Question 8 (EXPRESSION - 5 pts)
Write one line of code to get the list of **unique values** of the column ```time```.

**Output should look like this**

array(['Dinner', 'Lunch'], dtype=object)

In [17]:
# Your answer/code here:
df.time.unique()

array(['Dinner', 'Lunch'], dtype=object)

### 📑 Question 9 (EXPRESSION - 5 pts)
Write one line of code to get the range of table size. 

Range means the **difference between the max and the min**.

**Output should be a number**

In [19]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [22]:
# Your answer/code here:
df['size'].max() - df['size'].min()

5

### 📑 Question 10 (MULTICHOICE_SINGLE - 5 pts)
How many men/women paid for the dish?

- A Male: 157, Female: 89

- B Male: 85, Female: 156

- **C Male: 157, Female: 87**

- D Male: 142, Female: 15

In [23]:
# Your answer/code here:
# option 1:
df.groupby('sex')['tip'].count()

sex
Female     87
Male      157
Name: tip, dtype: int64

In [24]:
# option 2:
df.sex.value_counts()

Male      157
Female     87
Name: sex, dtype: int64

### 📑 Question 11 (MULTICHOICE_SINGLE - 5 pts)
What is the percentage of number of smoker table over the total?

- **A 38%**

- B 40%

- C 2%

- D 62%

In [27]:
# Your answer/code here:
len(df[df['smoker'] == 'Yes'])/len(df)

0.38114754098360654

In [31]:
df[df['smoker'] == 'Yes'].shape[0] / df.shape[0]

0.38114754098360654

In [37]:
sum(df['smoker'] == 'Yes') / len(df)

0.38114754098360654

### 📑 Question 12 (MULTICHOICE_SINGLE - 5 pts)
What is the percentage of revenue from the smoker tables over the total?

- A 38%

- **B 40%**

- C 2%

- D 62%

In [41]:
# Your answer/code here:
# option 1:
df[df.smoker == 'Yes']['total_bill'].sum() / df.total_bill.sum()

0.3998409203421041

In [44]:
# option 2:
df.groupby('smoker')['total_bill'].sum()/df.total_bill.sum()

smoker
No     0.600159
Yes    0.399841
Name: total_bill, dtype: float64

### 📑 Question 14 (MULTICHOICE_SINGLE - 5 pts)
Between smoker and non-smoker, who tips more according the percentage of the total tip on total bill.

- **A Non-smoker**

- B Smoker

In [50]:
# Your answer/code here:
df2 = df.groupby('smoker')['tip', 'total_bill'].sum().reset_index()
df2

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_8716\3299765213.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df2 = df.groupby('smoker')['tip', 'total_bill'].sum().reset_index()


smoker     tip  total_bill
0     No  451.77     2897.43
1    Yes  279.81     1930.34

In [52]:
df2['percentage_tip_totalbill'] = df2.tip / df2.total_bill
df2

smoker     tip  total_bill  percentage_tip_totalbill
0     No  451.77     2897.43                  0.155921
1    Yes  279.81     1930.34                  0.144954

### 📑 Question 15 (EXPRESSION - 5 pts)

Write one line of code to filter out the **"smoker"** tables on **Sunday**.

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 164 |        17.51 |  3    | Female | Yes      | Sun   | Dinner |      2 |
| 172 |         7.25 |  5.15 | Male   | Yes      | Sun   | Dinner |      2 |
| 173 |        31.85 |  3.18 | Male   | Yes      | Sun   | Dinner |      2 |
| 174 |        16.82 |  4    | Male   | Yes      | Sun   | Dinner |      2 |
| 175 |        32.9  |  3.11 | Male   | Yes      | Sun   | Dinner |      2 |
| 176 |        17.89 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 177 |        14.48 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 178 |         9.6  |  4    | Female | Yes      | Sun   | Dinner |      2 |
| 179 |        34.63 |  3.55 | Male   | Yes      | Sun   | Dinner |      2 |
| 180 |        34.65 |  3.68 | Male   | Yes      | Sun   | Dinner |      4 |
| 181 |        23.33 |  5.65 | Male   | Yes      | Sun   | Dinner |      2 |
| 182 |        45.35 |  3.5  | Male   | Yes      | Sun   | Dinner |      3 |
| 183 |        23.17 |  6.5  | Male   | Yes      | Sun   | Dinner |      4 |
| 184 |        40.55 |  3    | Male   | Yes      | Sun   | Dinner |      2 |
| 186 |        20.9  |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 187 |        30.46 |  2    | Male   | Yes      | Sun   | Dinner |      5 |
| 188 |        18.15 |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 189 |        23.1  |  4    | Male   | Yes      | Sun   | Dinner |      3 |
| 190 |        15.69 |  1.5  | Male   | Yes      | Sun   | Dinner |      2 |

In [53]:
# Your answer/code here:
df[(df.smoker == 'Yes') & (df.day == 'Sun')]

total_bill   tip     sex smoker  day    time  size
164       17.51  3.00  Female    Yes  Sun  Dinner     2
172        7.25  5.15    Male    Yes  Sun  Dinner     2
173       31.85  3.18    Male    Yes  Sun  Dinner     2
174       16.82  4.00    Male    Yes  Sun  Dinner     2
175       32.90  3.11    Male    Yes  Sun  Dinner     2
176       17.89  2.00    Male    Yes  Sun  Dinner     2
177       14.48  2.00    Male    Yes  Sun  Dinner     2
178        9.60  4.00  Female    Yes  Sun  Dinner     2
179       34.63  3.55    Male    Yes  Sun  Dinner     2
180       34.65  3.68    Male    Yes  Sun  Dinner     4
181       23.33  5.65    Male    Yes  Sun  Dinner     2
182       45.35  3.50    Male    Yes  Sun  Dinner     3
183       23.17  6.50    Male    Yes  Sun  Dinner     4
184       40.55  3.00    Male    Yes  Sun  Dinner     2
186       20.90  3.50  Female    Yes  Sun  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
188       18.15  3.50  Female    Yes  Sun  Dinner     3
189       23.10  4.00    Male    Yes  Sun  Dinner     3
190       15.69  1.50    Male    Yes  Sun  Dinner     2

### 📑 Question 16 (MULTICHOICE_SINGLE - 5 pts)
How many tables belong to the smoker, at lunch?

- A 20

- B 21 

- C 22

- **D 23**

In [55]:
# Your answer/code here:
df[(df.smoker == 'Yes') & (df.time == 'Lunch')].shape

(23, 7)

### 📑 Question 17 (EXPRESSION - 10 pts)

Write one line of code to compute the **average tip** amount of men who smoke and dine on **Friday's dinner**.


**Output should be a number**

In [59]:
# Your answer/code here:
df[(df.sex == 'Male') & (df.smoker == 'Yes') & (df.time == 'Dinner') & (df.day == 'Fri')]['tip'].mean()

3.246

### 📑 Question 18 (EXPRESSION - 5 pts)
Write one line of code to compute the **sum** of total bills of the **male** customers who dine in any day that's **not Friday**. 

**Output should be a number**

In [60]:
# Your answer/code here:
df[(df.sex == 'Male') & (df.time == 'Dinner') & (df.day != 'Fri')]['total_bill'].sum()


2496.81

### 📑 Question 19 (EXPRESSION - 5 pts)
Write one line of code to filter the tables with tip amount **larger** than the **average** tip amount. 

In [61]:
# Your answer/code here:
df.tip.mean()

2.9982786885245902

In [62]:
df[df.tip > df.tip.mean()]

total_bill   tip     sex smoker   day    time  size
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
5         25.29  4.71    Male     No   Sun  Dinner     4
7         26.88  3.12    Male     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
232       11.61  3.39    Male     No   Sat  Dinner     2
234       15.53  3.00    Male    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
239       29.03  5.92    Male     No   Sat  Dinner     3
243       18.78  3.00  Female     No  Thur  Dinner     2

[121 rows x 7 columns]

### 📑 Question 20 (EXPRESSION - 5 pts)

- Task 1: load `time_hours.xlsx` table to `df_th` variable:

In [63]:
df_th = pd.read_excel('data\\time_hours.xlsx') # your code here

# show time_hours table:
df_th

time    hour
0  Dinner  19-21h
1   Lunch  10-12h

- Task 2: Get `hour` information for `df` dataframe from `df_th` by using Using `time` field (`pd.merge()`):

In [65]:
# your code here:
df = pd.merge(
    left = df,
    right = df_th,
    on = 'time',
    how = 'left'
)

In [67]:
# show random sample 5 rows:
df.sample(5)

total_bill   tip     sex smoker   day    time  size    hour
138       16.00  2.00    Male    Yes  Thur   Lunch     2  10-12h
58        11.24  1.76    Male    Yes   Sat  Dinner     2  19-21h
183       23.17  6.50    Male    Yes   Sun  Dinner     4  19-21h
70        12.02  1.97    Male     No   Sat  Dinner     2  19-21h
197       43.11  5.00  Female    Yes  Thur   Lunch     4  10-12h